Update README with basic usage instructions

kkirsche opened the pull request:
This pull request adds / begins to add instructions on how to use the tool so that it is more accessible to new users

crhym3 commented:
Thanks for this. `goacme` already contains some help text. How about making it part of `goacme help`.

kkirsche commented:
The reasoning for adding this to the readme was that at least for when I was interested in using it I had a very hard time convincing people to even download it because there wasn't a clear representaiton of what the workflow would look like. I do think it could be added to the help text, but feel it greatly benefits the project to have it at the top level to tell people what they are getting into and whether the tool fits their needs

kkirsche commented:
Sorry for the follow up, so with that said — would you still like this moved fully into help instead, or potentially do you think it'd be possible to find some kind of happy medium between the two options as I certainly understand your point about it being useful with the command itself?

crhym3 commented:
@kkirsche sorry for late reply. ok yeah, that makes sense but wanted to note a couple things:

We imported this into [golang.org/x/crypto/acme/internal](https://godoc.org/golang.org/x/crypto/acme/internal/acme) which, as the name says, isn't exported yet but will be when it's more polished. Until then I wanted this package to mirror the internal version.

The golang/x/ doesn't contain the `goacme` command line tool, so it will live here unconditionally, at least for now.

crhym3 commented on README.md:
```diff
@@ -13,3 +13,31 @@ Contents of this repo:
+
+## Usage
+
+First, create an unencrypted private key for use with account registration. This can be done using OpenSSL:
```
this step as completely optional as goacme can generate an account key for you and store it on disk.

kkirsche replied:
Hm, interesting. I couldn't figure out how to get it to do that for me, it kept saying I didn't have a key. Would you be able to share the equivalent way that `goacme` does this? I'll happily set this as optional or explain how it can be done via `goacme` to keep things more self-contained.

crhym3 replied:
Sure thing. It should work with `goacme reg -gen mailto:x@example.com`.
I had some reasoning behind `-gen` being false by default but happy to hear alternatives. Maybe `-gen` could be true by default.

```
$ goacme help reg
usage: goacme reg [-c config] [-gen] [-d url] [contact [contact ...]]

Reg creates a new account at a CA using the discovery URL
specified with -d argument. The default value is letsencrypt.
For more information about the discovery run goacme help disco.

Upon successful registration, a new config will be written to account.json
in the directory specified with -c argument. Default location of the config dir
is /home/crhyme/.config/acme.
If the config dir does not exist, it will be created.

Contact arguments can be anything: email, phone number, etc.

If -gen flag is not specified, and an account key does not exist, the command
will exit with an error.
```

kkirsche commented:
@crhym3 I've updated the pull request to attempt to address the points discussed. Please let me know if you feel this is a good fit or if it could still benefit from additional tweaks or revisions.

crhym3 commented on README.md:
````diff
@@ -13,3 +13,35 @@ Contents of this repo:
+goacme update -accept
+```
+
+Now, we are ready to request our certificate. To do this, we need a private key for the domain first.
````
generating cert key with openssl is also optional. goacme can do it, too.

crhym3 commented on README.md:
````diff
@@ -13,3 +13,35 @@ Contents of this repo:
+
+```
+openssl -out ~/.config/acme/www.domain.com.key 2048
+goacme cert www.domain.com
````
in fact, this won't use the key generated on the line above. to use manually generated key you'll need to:

    goacme cert -k path/to/example.com.key example.com

crhym3 replied:
also, a nit: do you mind using `example.com` or `example.org` instead of `www.domain.com`?

kkirsche replied:
I don't mind changing the cert part, this just means I misunderstood what the help text meant:

    The certificate will be placed alongside key file, specified with -k argument.
    If the key file does not exist, a new one will be created.
    Default location for the key file is {{.ConfigDir}}/domain.key,
    where domain is the actually domain name provided as the command argument.

I assumed this meant that it would use the key generated above since the domain.key would replace domain with example.com as I understood it.

crhym3 replied:
Oh, actually you're right. I missed the fact that default for `-k` is exactly the expected location in this example, but it's a happy coincidence: if you name the key file in a different way, it won't be used.

Could you add that to the readme? Stating the fact that manually generating a cert key is optional and goacme can also do it automatically.

kkirsche replied:
(Easier than committing and reviewing that way) I'll modify it to read as follows, does this sound reasonable?

Original:
   
    Now, we are ready to request our certificate. To do this, we need a private key for the domain first.

    openssl -out ~/.config/acme/example.com.key 2048
    goacme cert -k path/to/example.com.key example.com

Revised:

    Now, we are ready to request our certificate. The certificate will be placed alongside key file, specified with -k argument. If the key file does not exist, a new one will be created. Default location for the key file is `{{.ConfigDir}}/domain.key`, where domain is the actually domain name provided as the command argument, in this, it would be `~/.config/acme/example.com.key`. If you would prefer to generate the key yourself, this may be done using openssl or a similar tool, and has been included / shown in the example below.

    openssl -out ~/.config/acme/example.com.key 2048
    goacme cert example.com

crhym3 replied:
`{{.ConfigDir}}/domain.key` is a bug in goacme help cert text :) it should've been expanded into `~/.config/acme` or something.

Why don't you follow the same pattern as for registration above, where you show both manual and auto separately?

kkirsche replied:
Ah, I didn't know that. I'll see if that's fixable after we get through this PR. Does this seem better (main paragraph not indented to make it easier to read):

Now, we are ready to request our certificate. The certificate will be placed alongside key file, specified with `-k` argument. If the key file does not exist, a new one will be created. Default location for the key file is `~/.config/acme/domain.key`, where domain is the actually domain name provided as the command argument, in this, it would be `~/.config/acme/example.com.key`. If you would prefer to generate the key yourself, this may be done using openssl or a similar tool, and has been included / shown in the example below.

    # Manually Generated Private Domain Key
    openssl -out ~/.config/acme/example.com.key 2048
    goacme cert example.com

    # Automatically Generated Private Domain Key
    goacme cert -k ~/.config/acme/example.com.key example.com

crhym3 replied:
sounds good. a few things:

1. did you mean "openssl genrsa -out ... 2048"? (missing genrsa argument);
same for account key above, in the registration phase.
2. for auto-generated, there's no need for -k ~/config/... argument.
3. I would remove the whole sentence mentioning key location. like you
said, it's a "getting started" short version. if users want to get more
details, they can always do that by installing goacme and executing "goacme
help cert".

On 9 June 2016 at 19:18, Kevin Kirsche <notifications@github.com> wrote:

> In README.md
> <https://github.com/google/goacme/pull/27#discussion_r66494030>:
>
> > +
> > +# Automatically Generated Account Key
> > +goacme reg -gen mailto:your-email-address@gmail.com
> > +```
> > +
> > +Next, review the terms and if you agree, you can accept them like so:
> > +
> > +```
> > +goacme update -accept
> > +```
> > +
> > +Now, we are ready to request our certificate. To do this, we need a private key for the domain first.
> > +
> > +```
> > +openssl -out ~/.config/acme/www.domain.com.key 2048
> > +goacme cert www.domain.com
>
> Ah, I didn't know that. I'll see if that's fixable after we get through
> this PR. Does this seem better (main paragraph not indented to make it
> easier to read):
>
> Now, we are ready to request our certificate. The certificate will be
> placed alongside key file, specified with -k argument. If the key file
> does not exist, a new one will be created. Default location for the key
> file is ~/.config/acme/domain.key, where domain is the actually domain
> name provided as the command argument, in this, it would be
> ~/.config/acme/example.com.key. If you would prefer to generate the key
> yourself, this may be done using openssl or a similar tool, and has been
> included / shown in the example below.
>
> # Manually Generated Private Domain Key
> openssl -out ~/.config/acme/example.com.key 2048
> goacme cert example.com
>
> # Automatically Generated Private Domain Key
> goacme cert -k ~/.config/acme/example.com.key example.com
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/goacme/pull/27/files/f99bc3fc5e8a3a1b822c8d7a0d866f57770c6195#r66494030>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABjPX7zqckkbsjcTIvpjGvZjT3EUKjCks5qKFj3gaJpZM4IsnSt>
> .
>

crhym3 replied:
forgot one more thing: `goacme cert` currently implements only `http-01` challenge, which requires the command to be run in a way the challenge can be resolved on the same machine, i.e. it's running a local HTTP server for the duration of authorization phase.

kkirsche replied:
True, I had to hack DNS to get that to work. I'll note that. Yes I did mean genrsa, that was a mistake.  Revised:

Now, we are ready to request our certificate. The certificate will be placed alongside key file, specified with the `-k` argument. If the key file does not exist, a new one will be created. If you would prefer to generate the key yourself, this may be done using openssl or a similar tool, and has been included / shown in the example below. `goacme cert` currently implements only the HTTP challenge mechanism, which requires the command to be run in a way the challenge can be resolved on the same machine, i.e. it's running a local HTTP server for the duration of authorization phase.

    # Manually Generated Private Domain Key
    openssl genrsa -out ~/.config/acme/example.com.key 2048
    goacme cert example.com

    # Automatically Generated Private Domain Key
    goacme cert example.com

crhym3 replied:
sounds good. do commit. remember to add genrsa to the account reg example,
too.

On 9 June 2016 at 20:01, Kevin Kirsche <notifications@github.com> wrote:

> In README.md
> <https://github.com/google/goacme/pull/27#discussion_r66501710>:
>
> > +
> > +# Automatically Generated Account Key
> > +goacme reg -gen mailto:your-email-address@gmail.com
> > +```
> > +
> > +Next, review the terms and if you agree, you can accept them like so:
> > +
> > +```
> > +goacme update -accept
> > +```
> > +
> > +Now, we are ready to request our certificate. To do this, we need a private key for the domain first.
> > +
> > +```
> > +openssl -out ~/.config/acme/www.domain.com.key 2048
> > +goacme cert www.domain.com
>
> True, I had to hack DNS to get that to work. I'll note that. Yes I did
> mean genrsa, that was a mistake. Revised:
>
> Now, we are ready to request our certificate. The certificate will be
> placed alongside key file, specified with the -k argument. If the key
> file does not exist, a new one will be created. If you would prefer to
> generate the key yourself, this may be done using openssl or a similar
> tool, and has been included / shown in the example below. goacme cert
> currently implements only the HTTP challenge mechanism, which requires the
> command to be run in a way the challenge can be resolved on the same
> machine, i.e. it's running a local HTTP server for the duration of
> authorization phase.
>
> # Manually Generated Private Domain Key
> openssl genrsa -out ~/.config/acme/example.com.key 2048
> goacme cert example.com
>
> # Automatically Generated Private Domain Key
> goacme cert example.com
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/goacme/pull/27/files/f99bc3fc5e8a3a1b822c8d7a0d866f57770c6195#r66501710>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABjPdmfq_GMNIod86IBPdIz8YFOvYNZks5qKGMLgaJpZM4IsnSt>
> .
>

crhym3 commented on README.md:
````diff
@@ -13,3 +13,35 @@ Contents of this repo:
+```
+mkdir -p ~/.config/acme
+openssl -out ~/.config/acme/account.key 2048
+```
````
a nit: could you remove all triple back-quotes and just intend the blocks with 4 spaces? It'll make the readme cleaner as there's no special syntax highlighting required.

kkirsche replied:
Not a nit at all, honestly I had forgotten that was even an option.

crhym3 commented on README.md:
```diff
@@ -13,3 +13,35 @@ Contents of this repo:
 ## License
```
Let's move the license to the bottom.

kkirsche commented:
Updated as discussed.

crhym3 commented:
Thanks. There were a couple other comments, e.g. https://github.com/google/goacme/pull/27#discussion_r66465573

kkirsche commented:
Ah, somehow I had missed those in my browser, sorry about. I'll look at those and take care of them

kkirsche commented:
Pushed those changes, did note where the confusion was on comment 27 though in relation to the help text.

kkirsche commented:
Sorry for the delay on committing this. Meetings and whatnot came up. This should now have all the changes discussed.

crhym3 commented on README.md:
```diff
@@ -10,6 +10,34 @@ Contents of this repo:
+
+    goacme update -accept
+
+    Now, we are ready to request our certificate. The certificate will be placed alongside key file, specified with the `-k` argument. If the key file does not exist, a new one will be created. If you would prefer to generate the key yourself, this may be done using openssl or a similar tool, and has been included / shown in the example below. `goacme cert` currently implements only the HTTP chal ...
```
Two last (I hope) things:

The line is very long, can you split it in around 80 chars each

"... requires the command to be run in the same machine" is not strictly true anymore. I added a `-manual` flag, see this for more info: https://github.com/google/goacme/commit/d14a85d62e085a883cb7d8ba03131dcd668e60bd

Essentially, it allows you to run `goacme cert` on a different host, as long as you have access to where the domain is served from.

kkirsche commented:
Shortened line lengths, adjusted the verbiage regarding the challenge, and added one example showing the manual challenge flag. Please let me know if there is anything that I missed

crhym3 commented:
Thanks for this. Squashing and merging now.

kkirsche commented:
Of course. Thanks for being patient and helpful with all the revisions
